compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -955,6 +960,94 @@ impl<'a> Parser<'a> {
         }
     }
 
+    /// Recover from `Fn`-family traits (Fn, FnMut, FnOnce) with lifetime arguments
+    /// (e.g. `FnOnce<'a>(&'a str) -> bool`). Up to generic arguments have already
+    /// been eaten.
+    fn recover_fn_trait_with_lifetime_params(
+        &mut self,
+        fn_path: &mut ast::Path,
+        lifetime_defs: &mut Vec<GenericParam>,
+    ) -> PResult<'a, ()> {
+        let Some(fn_path_segment) = fn_path.segments.last_mut() else {
+            unreachable!();
+        };
+
+        let generic_args = if let Some(p_args) = &fn_path_segment.args {
+            p_args.clone().into_inner()
+        } else {
+            // Normally it wouldn't come here because the upstream should have parsed
+            // generic parameters (otherwise it's impossible to call this function).
+            return Ok(());
+        };
+        let lifetimes =
+            if let ast::GenericArgs::AngleBracketed(ast::AngleBracketedArgs { span: _, args }) =
+                &generic_args
+            {
+                args.into_iter()
+                    .filter_map(|arg| {
+                        if let ast::AngleBracketedArg::Arg(generic_arg) = arg
+                            && let ast::GenericArg::Lifetime(lifetime) = generic_arg {
+                            Some(lifetime)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect()
+            } else {
+                Vec::new()
+            };
+        // Only try to recover if the trait has lifetime params.
+        if lifetimes.is_empty() {
+            return Ok(());
+        }
+
+        // Parse `(T, U) -> R`.
```

Comment:
Couldn't you replace all of the following with `self.parse_fn_decl(…)` or am I missing something?

Reply:
Thanks, I've tried. But I think it'll be more difficult to get `inputs_span` when using `parse_fn_decl`:

```rust
// Here, `inputs` refer to their own spans, but do not include "(" and ")".
let ast::FnDecl { inputs, output } = self.parse_fn_decl(...)?.into_inner();
```

Any better way here?

Reply:
What you can do instead then is to take `self.token.span` or `self.prev_token.span` and create a new one (`lo.to(hi)`) that does cover the parens, and rely on the existing parsing logic. Alternatively, you might want to change `inputs` to include the parens (and check the test suite to see if that causes any regressions).

Reply:
@estebank 
Sorry, I'm getting a little confused.

> What you can do instead then is to take self.token.span or self.prev_token.span and create a new one (lo.to(hi)) that does cover the parens,

Given `Fn<'a>(&'a str) -> bool`, for example, we're trying to get the span referring to `(&'a str)` to use as the `inputs_span` in the `ast::ParenthesizedArgs`. But `parse_fn_decl()` eats `(&'a str) -> bool`, so `self.prev_token.span` refers to `bool` after that.

> Alternatively, you might want to change inputs to include the parens

I'm not sure I understand this. The type of the `inputs` field in the `FnDecl` struct, which is returned from `parse_fn_decl()`, is a `Vec` of input params. So I think it's difficult to change `inputs` to include the parens.

Reply:
I'm wondering if we change the visibility of `parse_fn_params` to `pub(super)` and use it to parse the input params here. It'll be more simple and readable than `self.parse_paren_comma_seq(|p| p.parse_ty())`, and may work well because easy to get `inputs_span`. I think this visibility change is consistent because `parse_ret_ty` has `pub(super)` visibility too.

Reply:
Yes, changing visibility in that way is reasonable

Reply:
Thanks! Changed.